### PR TITLE
Web3 rework + misc fixes.

### PR DESCRIPTION
**Pull request by hackmod**
- [x] restore accidentally removed web3 batch-job code to make it work as intended.
  - without batch job, geth/multi-geth does not work as intended.

(please see commit 899b5ff6c4)

- restore various accidentally removed features
  - [x] restore `customCss` (accidentally removed by commit f281d915)
  - [x] restore `showRichList` option
  - [x] use web3 2.0.0-alpha.1
  - [x] replace the outdated web3-explorer with the new web3 2.x method.

## Notes
- low-level batch job method only works with the web3 up to version 1.2.1.
- ~~websocket seems not stable enough~~

### CODE REVIEW

**Comment by hackmod** [comment]
web3 1.2.1 websocket seems not stable enough
~~~
<--- Last few GCs --->

[16217:0x3d4a070]  3372700 ms: Mark-sweep 2343.8 (2624.6) -> 2343.6 (2624.6) MB, 3307.0 / 0.0 ms  allocation failure GC in old space requested
[16217:0x3d4a070]  3376665 ms: Mark-sweep 2343.6 (2624.6) -> 2343.5 (2583.6) MB, 3964.2 / 0.0 ms  last resort GC in old space requested
[16217:0x3d4a070]  3379630 ms: Mark-sweep 2343.5 (2583.6) -> 2343.5 (2576.6) MB, 2964.0 / 0.0 ms  last resort GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x2d448b125891 <JSObject>
    1: _addResponseCallback [/home/explorer/node_modules/web3-providers-ws/src/index.js:~227] [pc=0x4540f7790f1](this=0x2b41c11cec1 <WebsocketProvider map = 0xdf48485d5e9>,payload=0x1fd4ec1d45e1 <Object
 map = 0xdf48484e1d9>,callback=0x1fd4ec1d4619 <JSFunction (sfi = 0xf2e565027d9)>)
    2: send [/home/explorer/node_modules/web3-providers-ws...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
...
~~~

File: tools/richlist.js [hunk 1]
```diff
@@ -423,27 +480,36 @@ function prepareJsonAddress(json, defaultType = 0) {
...
+  let batch = new web3.BatchRequest();
+  // normally, the request size of batch getBalance() of all accounts is not bigger than 128kB.
+  // simply getBalance at once using rpc batch job.
+  for (let i = 0; i < accounts.length; i++) {
+    let account = accounts[i];
+    batch.add(web3.eth.getBalance.request(account));
+  }
+
+  try {
+    // web3 2.0.0 case
```
Comment (streetcrypto7) [inline tools/richlist.js]: By using web3 2.0, you should remove web3explorer(web3);

Comment (streetcrypto7) [reply tools/richlist.js]: Same with this PR https://github.com/ethereumclassic/explorer/pull/328, you should also replace web3-explorer package which is used for parity specific RPC calls.

Comment (streetcrypto7) [reply tools/richlist.js]: Ref: https://web3js.readthedocs.io/en/v1.2.1/web3.html#extend

**Comment by streetcrypto7** [comment]
Could you please merge this pr https://github.com/ethereumclassic/explorer/pull/328 into this PR if not bother? Thanks.

File: tools/sync.js [hunk 2]
```diff
@@ -143,7 +143,7 @@ const writeTransactionsToDB = async (config, blockData, flush) => {
     self.miners = [];
   }
   if (blockData) {
-    self.miners.push({ address: blockData.miner, blockNumber: blockData.number, type: 0 });
+    self.miners.push({ address: blockData.miner.toLowerCase(), blockNumber: blockData.number, type: 0 });
```
Comment (streetcrypto7) [inline tools/sync.js]: No batch jobs for web3.eth.getTransactionReceipt(txData.hash)?

Comment (streetcrypto7) [reply tools/sync.js]: And why not make a PR to upstream web3 module to add parity methods?

File: tools/richlist.js [hunk 3]
```diff
@@ -477,4 +543,17 @@ async function startSync() {
     makeRichList(latestBlock, 500, updateAccounts);
   }
 }
-startSync();
+
+web3.eth.getNodeInfo((err, nodeInfo) => {
+  let isParity = false;
+
+  console.log(`Node version = ${nodeInfo}`);
+  if (nodeInfo.split('/')[0].toLowerCase().includes('parity')) {
```
Comment (streetcrypto7) [inline tools/richlist.js]:
```TypeError: Cannot read property 'split' of null
    at GetNodeInfoMethod.web3.eth.getNodeInfo (/home/ubuntu/explorer/routes/web3relay.js:54:16)
    at GetNodeInfoMethod._callee$ (/home/ubuntu/explorer/node_modules/web3-core-method/dist/web3-core-method.cjs.js:126:22)
    at tryCatch (/home/ubuntu/explorer/node_modules/regenerator-runtime/runtime.js:45:40)
    at Generator.invoke [as _invoke] (/home/ubuntu/explorer/node_modules/regenerator-runtime/runtime.js:271:22)
    at Generator.prototype.(anonymous function) [as next] (/home/ubuntu/explorer/node_modules/regenerator-runtime/runtime.js:97:21)
    at asyncGeneratorStep (/home/ubuntu/explorer/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
    at _next (/home/ubuntu/explorer/node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
    at process._tickCallback (internal/process/next_tick.js:68:7)
TypeError: Cannot read property 'split' of null
    at GetNodeInfoMethod.web3.eth.getNodeInfo (/home/ubuntu/explorer/routes/web3relay.js:54:16)
    at GetNodeInfoMethod._callee$ (/home/ubuntu/explorer/node_modules/web3-core-method/dist/web3-core-method.cjs.js:126:22)
    at tryCatch (/home/ubuntu/explorer/node_modules/regenerator-runtime/runtime.js:45:40)
    at Generator.invoke [as _invoke] (/home/ubuntu/explorer/node_modules/regenerator-runtime/runtime.js:271:22)
    at Generator.prototype.(anonymous function) [as next] (/home/ubuntu/explorer/node_modules/regenerator-runtime/runtime.js:97:21)
    at asyncGeneratorStep (/home/ubuntu/explorer/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
    at _next (/home/ubuntu/explorer/node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

Could you see the documentation please? https://web3js.readthedocs.io/en/v2.0.0-alpha.1/web3-eth.html

Comment (streetcrypto7) [reply tools/richlist.js]: More logs

```
Web3 has detected parity node configuration
Web3 has detected parity node configuration
{ Error: Cannot find module '../../methods/TraceFilterMethod'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/ubuntu/explorer/lib/trace/factories/MethodFactory.js:10:1)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Module._compile (/home/ubuntu/explorer/node_modules/pirates/lib/index.js:99:24)
    at Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Object.newLoader [as .js] (/home/ubuntu/explorer/node_modules/pirates/lib/index.js:104:7)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/ubuntu/explorer/lib/trace/index.js:13:1)
    at Module._compile (internal/modules/cjs/loader.js:778:30) code: 'MODULE_NOT_FOUND' }
{ Error: Cannot find module '../../methods/TraceFilterMethod'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/ubuntu/explorer/lib/trace/factories/MethodFactory.js:10:1)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Module._compile (/home/ubuntu/explorer/node_modules/pirates/lib/index.js:99:24)
    at Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Object.newLoader [as .js] (/home/ubuntu/explorer/node_modules/pirates/lib/index.js:104:7)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/ubuntu/explorer/lib/trace/index.js:13:1)
    at Module._compile (internal/modules/cjs/loader.js:778:30) code: 'MODULE_NOT_FOUND' }

/home/ubuntu/explorer/node_modules/solc/soljson.js:33
(Module.asmGlobalArg,Module.asmLibraryArg,buffer);var __GLOBAL__I_000101=Module["__GLOBAL__I_000101"]=asm["__GLOBAL__I_000101"];var __GLOBAL__sub_I_ABIFunctions_cpp=Module["__GLOBAL__sub_I_ABIFunctions_cpp"]=asm["__GLOBAL__sub_I_ABIFunctions_cpp"];var __GLOBAL__sub_I_ABI_cpp=Module["__GLOBAL__sub_I_ABI_cpp"]=asm["__GLOBAL__sub_I_ABI_cpp"];var __GLOBAL__sub_I_ASTJsonConverter_cpp=Module["__GLOBAL__sub_I_ASTJsonConverter_cpp"]=asm["__GLOBAL__sub_I_ASTJsonConverter_cpp"];var __GLOBAL__sub_I_AST_cpp=Module["__GLOBAL__sub_I_AST_cpp"]=asm["__GLOBAL__sub_I_AST_cpp"];var __GLOBAL__sub_I_ArrayUtils_cpp=Module["__GLOBAL__sub_I_ArrayUtils_cpp"]=asm["__GLOBAL__sub_I_ArrayUtils_cpp"];var __GLOBAL__sub_I_AsmAnalysis_cpp=Module["__GLOBAL__sub_I_AsmAnalysis_cpp"]=asm["__GLOBAL__sub_I_AsmAnalysis_cpp"];var __GLOBAL__sub_I_BMC_cpp=Module["__GLOBAL__sub_I_BMC_cpp"]=asm["__GLOBAL__sub_I_BMC_cpp"];var __GLOBAL__sub_I_CHC_cpp=Module["__GLOBAL__sub_I_CHC_cpp"]=asm["__GLOBAL__s
abort({"code":"MODULE_NOT_FOUND"}). Build with -s ASSERTIONS=1 for more info.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! etc-explorer@0.2.0 app: `node app.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the etc-explorer@0.2.0 app script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/ubuntu/.npm/_logs/2019-10-20T13_38_55_565Z-debug.log
```

Comment (streetcrypto7) [reply tools/richlist.js]: Parity lib doesn't work either

```
(node:2209) UnhandledPromiseRejectionWarning: Error: Cannot find module '../../methods/ParityListAccountsMethod'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/ubuntu/explorer/lib/parity/factories/MethodFactory.js:10:1)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Module._compile (/home/ubuntu/explorer/node_modules/pirates/lib/index.js:99:24)
    at Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Object.newLoader [as .js] (/home/ubuntu/explorer/node_modules/pirates/lib/index.js:104:7)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/ubuntu/explorer/lib/parity/index.js:13:1)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
(node:2209) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:2209) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

Comment (hackmod) [reply tools/richlist.js]:
~~~
import {Network} from 'web3-net';
       ^

SyntaxError: Unexpected token {
~~~
the latest web3 extension uses the babel-* packages.
I guess you've missed `npm i`

Comment (streetcrypto7) [reply tools/richlist.js]: @hackmod I did `npm install` and I am using the latest LTS version of nodejs which is node 10

Comment (hackmod) [reply tools/richlist.js]: thanks for your patience.
I've missed some `lib/methods` files that caused an error and now fixed.
thanks!

**Comment by kimisan** [comment]
@hackmod 
may you check the continuous-integration/travis-ci/pr ?

**Comment by streetcrypto7** [comment]
@kimisan It is because of docker's nodejs version

https://github.com/ethereumclassic/explorer/blob/master/Dockerfile#L1

This can be fixed to `FROM node:10`

**Comment by streetcrypto7** [comment]
And you should update package-lock.json file as well,

see mine https://travis-ci.org/streetcrypto7/explorer/builds/600717902?utm_source=github_status&utm_medium=notification

**Comment by hackmod** [comment]
@kimisan // node version issue resolved. ready to merge now!

**Comment by hackmod** [comment]
image
travis-ci passed at https://travis-ci.org/EthersocialNetwork/explorer/builds/625344321

**Comment by kimisan** [comment]
@phyro @realcodywburns 
Can you help to merge the pull request.
Thanks

**Comment by realcodywburns** [comment]
Merging w/o travis ci test run.need to troubleshoot ci/cd pipeline. Code passed in abo e mentioned tests

**Pull request merged**